add ICC profiles to PDFs

eroux opened the pull request:
Fixes #697

- [ ] The GitHub pipeline is OK (green), <!-- The maintainers will trigger it if this is your 1st contribution -->
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [X] A unit test is covering the code added / modified by this PR

- [X] This PR is ready to be merged <!-- In your opinion, can this be merged as soon as it's reviewed? Else, this can be turned into a Draft PR -->

- [ ] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [X] A mention of the change is present in `CHANGELOG.md`

By submitting this pull request, I confirm that my contribution is made under the terms of the [GNU LGPL 3.0 license](https://github.com/PyFPDF/fpdf2/blob/master/LICENSE).

eroux commented:
@Lucas-C I don't have a Winows machine so I couldn't update `test/image/image_types/image_types_insert_png_alpha_dctdecode_windows.pdf`, is there any way you could do it? Thanks!

Lucas-C commented on fpdf/output.py:
```diff
@@ -194,7 +194,6 @@ class PDFXObject(PDFContentStream):
         "height",
         "color_space",
         "bits_per_component",
-        "filter",
```
👍 Well spotted!

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -3719,6 +3720,15 @@ def preload_image(self, name, dims=None):
             info["i"] = len(self.images) + 1
             info["usages"] = 1
+            info["iccp_i"] = None
+            if "iccp" in info and info["iccp"]:
+                if info["iccp"] in self.iccps:
+                    info["iccp_i"] = self.iccps[info["iccp"]]
+                else:
+                    iccp_i = len(self.iccps)
+                    self.iccps[info["iccp"]] = iccp_i
+                    info["iccp_i"] = iccp_i
+                info["iccp"] = None
```
```suggestion
            iccp = info.get("iccp")
            if iccp:
                if iccp in self.iccps:
                    info["iccp_i"] = self.iccps[iccp]
                else:
                    iccp_i = len(self.iccps)
                    self.iccps[iccp] = iccp_i
                    info["iccp_i"] = iccp_i
                info["iccp"] = None
```

Lucas-C commented on fpdf/output.py:
```diff
@@ -719,10 +740,33 @@ def _add_images(self):
+        Returns the PDF object of the ICC profile indexed iccp_i in the FPDF object.
+        Adds it if not present.
+        """
+        if iccp_i in self.iccp_i_to_pdf_i:
```
```suggestion
        iccp_i = img_info["iccp_i"]
        if iccp_i in self.iccp_i_to_pdf_i:
```

Lucas-C commented on test/image/test_load_image.py:
```diff
@@ -45,10 +45,12 @@ def test_load_invalid_base64_data():
     def build_pdf_with_big_images():
         pdf = fpdf.FPDF()
         pdf.images = images_cache
+        pdf.iccps = iccp_cache
```
👍

Lucas-C commented on test/image/image_types/test_insert_images.py:
```diff
@@ -116,6 +116,14 @@ def test_insert_bmp(tmp_path):
     assert_pdf_equal(pdf, HERE / "image_types_insert_bmp.pdf", tmp_path)
 
 
+def test_insert_jpg_icc(tmp_path):
```
Could you please add a unit test when the same image is added twice in the same doc?
I'd like to ensure that a single `PDFICCPObject` object gets inserted in that case 😊

eroux replied:
done, I also added the case where the same ICC profile is included in multiple images

Lucas-C commented on fpdf/image_parsing.py:
```diff
@@ -353,6 +353,10 @@ def get_img_info(filename, img=None, image_filter="AUTO", dims=None):
 
+    iccp = None
+    if "icc_profile" in img.info:
+        iccp = img.info.get("icc_profile")
```
Given that this object is never altered by your code later on,
that means that the `icc_profile` provided by Pillow has the right format for the PDF spec,
no need for any data manipulation? If so that's great and very handy! 😊

eroux replied:
yes, the ICC profile doesn't need to be parsed or changed

For the validation... I guess I could add something but I'm not sure how useful that might be. img2pdf is usually very thorough in checking everything and it doesn't check this so my guess is that it must be quite uncommon In any case, Pillow depends on [pycms](https://www.cazabon.com/pyCMS/downloads/index.txt) which has a `getProfileInfo`, I guess I could use that

eroux replied:
do you think it's really necessary? there's a chance that the image won't be read by PDF readers anyway

Lucas-C replied:
I suggest that you give this a try, as it would be nice to report potential issues regarding ICC profiles to `fpdf2` users,
but I agree that this should relatively uncommon 😊
In fact, my understand is that `Pillow` will raise an error for invalid ICC profiles: https://github.com/python-pillow/Pillow/blob/main/Tests/test_imagecms.py

eroux replied:
done (I think)

Lucas-C reviewed (changes requested):
Hey @eroux
Thank you for taking the time to work on this!
That's a very valuable addition for `fpdf2` 🥳😀

Lucas-C commented on fpdf/fpdf.py:
```diff
@@ -3719,6 +3720,15 @@ def preload_image(self, name, dims=None):
             info["usages"] = 1
+            info["iccp_i"] = None
+            if "iccp" in info and info["iccp"]:
+                if info["iccp"] in self.iccps:
```
Also, could you add a call like this here please:
```
LOGGER.debug("ICC profile found for image %s - It will be inserted in the PDF document", name)
```

Lucas-C commented on fpdf/output.py:
```diff
@@ -719,10 +740,33 @@ def _add_images(self):
         color_space = Name(info["cs"])
         decode = None
-        if color_space == "Indexed":
+        if "iccp_i" in info and info["iccp_i"] is not None:
```
```suggestion
        iccp_i = info.get("iccp_i")
        if iccp_i is not None:
```

eroux commented on fpdf/output.py:
```diff
@@ -719,10 +740,33 @@ def _add_images(self):
+        if "iccp_i" in info and info["iccp_i"] is not None:
+            iccp_pdf_i = self._ensure_iccp(info, info["iccp_i"])
+            color_space = PDFArray(["/ICCBased", str(iccp_pdf_i), str("0"), "R"])
+        elif color_space == "Indexed":
```
img2pdf raises an exception in that case, what do you think would be best?

Lucas-C replied:
I'd say, if `color_space == "Indexed"` and there is an ICC profile, ignore the ICC profile and log a warning about the fact that `fpdf2` does not support this use case for now. This way, we are sure to stay backward-compatible and not trigger any unwanted changes for `fpdf2` users processing potential indexed images with an ICC profile

eroux replied:
done

Lucas-C commented:
> @Lucas-C I don't have a Winows machine so I couldn't update `test/image/image_types/image_types_insert_png_alpha_dctdecode_windows.pdf`, is there any way you could do it? Thanks!

Sure, but I just ran unit test `test_insert_png_alpha_dctdecode()` on my Windows environment, and it passed!

By the way, you may want to get rid of `PYTHONMALLOCSTATS` as part of thi PR to avoid the GitHub Actions logs being bloated, _cf._ https://github.com/PyFPDF/fpdf2/pull/657#issuecomment-1444607110

eroux commented on fpdf/fpdf.py:
```diff
@@ -291,6 +291,7 @@ def __init__(
         self.pages = {}  # array of PDFPage objects starting at index 1
         self.fonts = {}  # map font string keys to dicts describing the fonts used
         self.images = {}  # map image identifiers to dicts describing the raster images
+        self.iccps = {}  # map icc profiles (bytes) to their index (number)
```
done

eroux commented on fpdf/output.py:
```diff
@@ -719,10 +740,33 @@ def _add_images(self):
         decode = None
-        if color_space == "Indexed":
+        if "iccp_i" in info and info["iccp_i"] is not None:
+            iccp_pdf_i = self._ensure_iccp(info, info["iccp_i"])
```
indeed! done

Lucas-C commented:
> Sure, but I just ran unit test `test_insert_png_alpha_dctdecode()` on my Windows environment, and it passed!

My bad, I did not run the test correctly... 🤦‍♂️

I generated a reference PDF file, but I got this error:
```
ERROR: Permission to buda-base/fpdf2.git denied to Lucas-C.
```
You can either allow me push to your fork: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

Or use the PDF file included with this comment:
[image_types_insert_png_alpha_dctdecode_windows.pdf](https://github.com/PyFPDF/fpdf2/files/10840541/image_types_insert_png_alpha_dctdecode_windows.pdf)

Lucas-C commented:
Apart from the nice mention you already added in `CHANGELOG.md`, I think this feature would be worth a section in `docs/images.md`! 😊

In order to mention this automatic inclusion of ICC profiles from images, how to remove /alter ICC profiles:
basically by modifying `.info["icc_profile"]` with Pillow before passing them to `FPDF.image()`

eroux commented:
thanks for the PDF, I've pushed it. I'm working on the two changes but I just wanted to say that updating `docs/images.md` is a good idea. 

I'm thinking also that perhaps releasing a new version with those changes could be a good option? The changes since the last version are pretty significant for use cases involving images

Lucas-C commented:
> I'm thinking also that perhaps releasing a new version with those changes could be a good option? The changes since the last version are pretty significant for use cases involving images

I agree.
Given the maturity of the 4 current open PRs, I'm currently planning to perform a release once they are all merged.
If ever any of them is delayed, I may publish a new version of `fpdf2` beforehand, we'll see.

Lucas-C commented on docs/Images.md:
```diff
@@ -150,8 +150,7 @@ pdf.image("https://upload.wikimedia.org/wikipedia/commons/7/70/Example.png")
 
-By default, `fpdf2` will avoid altering your images :
-no image conversion from / to PNG / JPEG is performed.
+By default, `fpdf2` will avoid altering or recompressing your images: when possible, the original bytes from the JPG or TIFF file will be used directly. Bitonal images are by default compressed as TIFF Group4.
```
👍

Lucas-C commented on test/image/image_types/test_insert_images.py:
```diff
@@ -116,6 +116,22 @@ def test_insert_bmp(tmp_path):
+    # only once in that case
+    pdf.add_page(format=(448, 498))
+    pdf.image(HERE / "insert_images_insert_jpg_icc.jpg", x=0, y=0, h=498)
+    # we add another image with the same ICC profile to make sure it's also included
```
👍 Perfect!

Lucas-C commented on fpdf/image_parsing.py:
```diff
@@ -311,6 +314,19 @@ def _decode_base64_image(base64Image):
 ]
 
 
+def iccp_is_valid(iccp):
```
Could you please add this unit test to cover that case?
```python
def test_insert_jpg_with_invalid_icc(caplog):
    pdf = fpdf.FPDF()
    pdf.add_page()
    pdf.image(HERE / "jpg_with_invalid_icc.jpg")
    assert "ICC Profile" in caplog.text
    assert "is invalid" in caplog.text
```

And here is the code to generate `jpg_with_invalid_icc.jpg`:
```python
from PIL import Image, ImageCms
img = Image.open(HERE / "insert_images_insert_jpg_icc.jpg")
iccp = img.info["icc_profile"]
img.save(HERE / "jpg_with_invalid_icc.jpg", icc_profile=iccp[5:])
```

eroux replied:
thanks! done

Lucas-C commented on test/image/image_types/test_insert_images.py:
```diff
@@ -116,6 +116,32 @@ def test_insert_bmp(tmp_path):
+    pdf.set_margin(0)
+    pdf.image(HERE / "insert_images_insert_jpg_icc_invalid.jpg", x=0, y=0, h=498)
+    assert_pdf_equal(
+        pdf, HERE / "image_types_insert_jpg_icc_invalid.pdf", tmp_path, generate=True
```
You forgot a `generate=True` there 😊

eroux replied:
ok yes, thanks! I've pushed a fix for the generate

Lucas-C replied:
Assertion added in #710

Lucas-C commented:
Merged!
Thank you @eroux for this contribution 👍

eroux commented:
amazing, thanks!
